notes/lsp-semantics.md

```diff
@@ -0,0 +1,146 @@
+also has a "find references" method, which is sort of like a goto.
+
+Here is the easy base for goto definition: when a variable is bound by a
+`let`, `fun`, or `match`, the variable binding (possibly in a pattern) is an LSP
+"definition" and any uses of that variable are LSP "references".
+
+For example, in `let foo = 3 in 4 + foo` a "list references" request on the
+first `foo` should return the second `foo`, and a "goto definition" on the
+second `foo` should return the first `foo`.
+
+Static accesses are a little more challenging: given a path like `foo.bar.baz`
+and a "goto definition" request for `baz`, it would be nice if we could find
+where the value for `baz` is set (i.e. some part of the code that looks like
+`{ ..., baz = 2, ...}`). This cannot be done in general (at least, without
+evaluation) because `foo` could be the result of some arbitrary computation.
+Nevertheless, we make some effort at static analysis for the common cases.
+
+Let's describe a few simple cases first:
+
+- The fields in a record literal
+  are LSP "definitions" and any static accesses of them are LSP "references".
+  For example, in `{bar = 3}.bar` the first occurrence of `bar` is
+  a definition and the second is a reference to it.
+
+- For all path elements except the last, `let` bindings should be transparent.
+  For example, in each of
+
+    - `let foo = { bar = 3 } in foo.bar`,
+    - `let baz = { bar = 3 } in let foo = baz in foo.bar`, and
+
+  the second occurrence of `bar` references the first. 
+
+- The definitions should be repeatedly "resolved" for all path elements
+  except the last.
+  In `let foo = { baz = { bar = 3 } } in foo.baz.bar`, the second occurrence
+  of `bar` references the first.
+
+- In the merge of two records, fields defined on both sides are definitions.
+  We don't take merge priorities into account, because it's probably useful for the user to
+  see the overridden values in addition to the values that "win." Note that the LSP allows
+  multiple responses to a "goto definition" request.
+
+  For example, in `let x = { foo | default = 3, bar = 4 } & { foo = 2 } in [x.foo, x.bar]`
+  the second instance of `bar` references the first instance, while the last instance of `foo`
+  references the first two instances.
+
+- If-then-else acts like a merge: both branches provide definitions, so that
+  in `let x = if ... then { foo = 1 } else { foo = 2 } in x.foo`, the last instance of `foo`
+  references the first two instances.
+
+- This might be getting too fancy, but we could potentially do static analysis of functions
+  to track the origin of record fields. For example, we might expect that sprinkling the
+  identity function around doesn't affect definition tracking.
+
+### The typescript-language-server
+
+`typescript-language-server` has a somewhat similar behavior to what we're aiming for, and
+it can even "see through" functions to some extent.
```

Comment:
This was what I was alluding to in the last vague bullet point. I've tried to write it a little more clearly now. One thing I'm not quite sure about is the extent to which we want to do beta-substitution here, because the record could be coming from the argument instead of the function body, like in `let f = fun x => x.foo in (f { foo = { bar = 1 } }).bar`.